### PR TITLE
Add SF.12, 3rd try

### PR DESCRIPTION
**Pull request by ChrisGuzak**

### CODE REVIEW

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -19181,6 +19182,33 @@ A header should include all its dependencies. Be careful about using relative pa
 
 A test should verify that the header file itself compiles or that a cpp file which only includes the header file compiles.
 
+### <a name="Rs-incscope"></a>SF.12: Prefer the angle bracket form of `#include` where you can and the quoted form everywhere else
+
+##### Reason
+
+The [standard](http://eel.is/c++draft/cpp.include) provides flexability for compilers to implement
+the two forms of `#include` selected using the angle (`<>`) or quoted (`""`) syntax. Vendors take
+advantage of this and and use different search algorythms and methods for specifying the include path.
```
Comment (shaneasd) [inline CppCoreGuidelines.md]: algorythms -> algorithms?

File: CppCoreGuidelines.md [hunk 2]
```diff
@@ -19181,6 +19182,33 @@ A header should include all its dependencies. Be careful about using relative pa
...
+
+Neveer the less, the gudance is to use the angle form when possible. This supports the fact that the
```
Comment (shaneasd) [inline CppCoreGuidelines.md]: ```Neveer the less, the gudance``` -> ```Nevertheless, the guidance```

File: CppCoreGuidelines.md [hunk 3]
```diff
@@ -19181,6 +19182,33 @@ A header should include all its dependencies. Be careful about using relative pa
...
+standard library headers must be includedthis way, is more likely to create portable code, and enables
```
Comment (shaneasd) [inline CppCoreGuidelines.md]: ```includedthis``` -> ```included this```

File: CppCoreGuidelines.md [hunk 4]
```diff
@@ -19181,6 +19182,33 @@ A header should include all its dependencies. Be careful about using relative pa
...
+the quoted form for other uses. For example being clear about the locality of the header relative
+to files that includes it or in scearios where the different search algorythm is required.
```
Comment (shaneasd) [inline CppCoreGuidelines.md]: scearios -> scenarios

**Comment by shaneasd** [comment]
I agree with what I believe to be the intended reading of this rule. Is it sufficiently clear, though, that if angle brackets don't work the solution is to use quotes rather than change search paths? i.e. in the scenario where ```<helpers.h>``` doesn't work I could fix this by either adding the directory in question to the includes search path or by using quotes.

Perhaps clarifying this would be too much in the direction of hand-holding but I can imagine this cropping up in a code review where a less experienced developer has interpreted literally as "where you can" rather than "where you can without breaking other implied rules"

File: CppCoreGuidelines.md [hunk 5]
```diff
@@ -19181,6 +19182,33 @@ A header should include all its dependencies. Be careful about using relative pa
...
+    #include "helpers.h"    // A project specific file, use "" form
+
+##### Note
+Failing to follow this results in difficult to diagnose errors due to picking up the wrong file by incorrectly specifying the scope when it is included.
+
+Library creators should put their headers in a folder and have clients include those files using the relative path `#include <some_library/common.h>`
+
+##### Enforcement
+
+A test should verify that the headers referenced via `<>` are from the standard library and those using `""` are not.
```
Comment (MikeGitb) [inline CppCoreGuidelines.md]: Doesn't that contradict the rule to use angle brackets where possible?

File: CppCoreGuidelines.md [hunk 6]
```diff
@@ -19181,6 +19182,33 @@ A header should include all its dependencies. Be careful about using relative pa
...
+
+##### Example
+
+    #include <string>       // From the standard library, required form
+    #include "helpers.h"    // A project specific file, use "" form
+
+##### Note
+Failing to follow this results in difficult to diagnose errors due to picking up the wrong file by incorrectly specifying the scope when it is included.
+
+Library creators should put their headers in a folder and have clients include those files using the relative path `#include <some_library/common.h>`
```
Comment (MikeGitb) [inline CppCoreGuidelines.md]: I think that point can't be stressed enough. The number of times I've seen example code that starts with `#include <json.h>`

File: scripts/hunspell/isocpp.dic [hunk 7]
```diff
@@ -244,6 +244,7 @@ incomplet
 incorrekt
 increment1
 Incrementable
+incscope
```
Comment (MikeGitb) [inline scripts/hunspell/isocpp.dic]: What is in scope?

File: CppCoreGuidelines.md (hunk 5 above)
Comment (ChrisGuzak) [reply CppCoreGuidelines.md]: I missed updating this, this was from the 2nd version.

File: scripts/hunspell/isocpp.dic (hunk 7 above)
Comment (ChrisGuzak) [reply scripts/hunspell/isocpp.dic]: the label in the markup, and I now see it is needs updating given the change in guidance.

**Comment by ChrisGuzak** [comment]
> ... in the scenario where `<helpers.h>` doesn't work I could fix this by either adding the directory in question to the includes search path or by using quotes.

good point, I can see how some will be tempted to do this. in fact in the OS build system "." is added to the include path making the <> form work in cases where normally it does not.

I'm open to edits that address this but maybe it does need to be.

File: CppCoreGuidelines.md (hunk 2 above)
Comment (shaneasd) [reply CppCoreGuidelines.md]: I'm reasonably sure it's ```Nevertheless``` not ```Never the less```

**Comment by shaneasd** [comment]
> I'm open to edits that address this but maybe it does need to be.

If I think of a way to express this concisely I'll suggest it but at the moment I'm struggling.

File: CppCoreGuidelines.md (hunk 2 above)
Comment (jwakely) [reply CppCoreGuidelines.md]: It is.

**Comment by hsutter** [comment]
Editors call: Great, thanks again for iterating on this and contributing it!

**Comment by johnmcfarlane** [comment]
Sorry for the infrequent feedback and thanks for putting the time into adding this rule.

The title could be read the right way if only the details were correct. However, the advise is the same as in #1456 and I still suspect it's not quite right. I know a lot of people follow this practice so that's understandable. However, it isn't aligned with what our common toolchains do so I'm still concerned that it will cause confusion and perpetuate bad habits.

> #include "helpers.h"    // A project specific file, use "" form

Project-specific files may need to use either `<>` or `""` depending on whether the user intends to locate *helpers.h* via a header search *within the project scope* or relative to the including file (respectively). The fact that the two files are within the confines of a single project is of no consequence.

> Failing to follow this results in difficult to diagnose errors due to picking up the wrong file by incorrectly specifying the scope when it is included.

This would be true if the advise were correct. But again, using `""` in an intra-project inclusion where `<>` is the correct thing to do is still going result in occasional errors. That's now worse if the user was following advise from the CCG.

Apart from anything else, how does this apply to libraries outside the project? They are projects too. Should they use `<>` or `""` when including their own headers? From their point of view, they're including project-specific headers, so we're saying they should use `""`. But from the point of view of a depenent project, they should be using `<>`. How does one resolve this contradiction if "being in a project" is the way we're distinguishing files?

> Library creators should put their headers in a folder and have clients include those files using the relative path `#include <some_library/common.h>`

The use of "relative" here is confusing because `""`-style inclusion is relative to the including file, whereas the `<>` style is absolute.

Please read the tool-chain documentation (***my emphasis***):

* [GCC](https://gcc.gnu.org/onlinedocs/cpp/Search-Path.html): "By default, the preprocessor looks for header files included by the quote form of the directive #include "file" first ***relative to the directory of the current file***"
* [MSVC](https://docs.microsoft.com/en-us/cpp/preprocessor/hash-include-directive-c-cpp?view=vs-2019): "The quotation marks mean that the preprocessor first searches ***the directory that contains the parent source file***."

Would anybody mind if I fix this or is everybody in disagreement with me? I ask because my [posts on the subject](https://github.com/isocpp/CppCoreGuidelines/pull/1456#issuecomment-565734871) in the previous PR didn't receive much feedback so I still don't know if I'm onto a losing battle here.

**Comment by shaneasd** [comment]
It seems to me like your objection is purely to ```// A project specific file, use "" form```? If so, I agree. It adds a new concept (project ownership) in the example where it should really be illustrating the established concepts and I agree with you that the new concept it introduces is not strictly appropriate.

**Pull request merged**